Ordered imports grouping

abierbaum opened the pull request:
#### PR checklist

- [x] Addresses an existing issue: helps #3450, allows for flexibility to address other custom cases
- [x] New enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:

This PR adds support for configuring custom groups with the ordered-import rule.

Previously the rule would only group into third-party, parent directories, and current directory.  This change allows for fully customizing the grouping but still keeps a default that provides exactly the same as was there previously.

An example configuration would be something like this:

```
"rules": {
    "ordered-imports": [
        true,
        {
            "import-sources-order": "case-insensitive",
            "named-imports-order": "case-insensitive",
            "grouped-imports": true,
            "groups": [
                { "match": "^@(rxjs/angular)", "order": 10 },
                { "match": "^app", "order": 20 },
                { "name": "parent_dir", "match": "^[.][.]", "order": 50 },
                { "name": "current dir", "match": "^[.]", "order": 40 },
                { "match": null, "order": 30 }
            ]
        }
    ]
}
```
This would enforce an ordering where all rxjs and angular imports would come in first group sorted, then all application imports, then all additional third-party imports, then relative parent imports, and finally current directory imports.

This provides the flexibility to support just many use cases and can be adapted easily.  

There are other projects such as [vscode import-sorter ](https://marketplace.visualstudio.com/items?itemName=mike-co.import-sorter) and a few unsupported tslint extensions that try to address this issue but it seems like this is something that would best be handled by tslint since it can already reorder imports.

One of the goals of this change is to allow using the fixer to automatically reformat imports on the fly.  So for example when using vscode or another editor with tslint support the user could enable the fixer to be run upon file save and have tslint restructure all imports to meet the given team's conventions.

#### Is there anything you'd like reviewers to focus on?
I tried to add some solid test examples but I may have missed a corner case.  Please let me know if you see something I should check.

I tried to keep the changes to a minimum, but I did refactor some of the existing code to make the grouping implementation a bit more clear and straight forward.  I also added some comments to clarify a few places in the code that tripped me up.  If you would like these stripped back out just let me know.

Note: There are a number of code style changes that were made by prettier as part of the precommit hook that is now part of tslint.  (just so you know I didn't go try to change the original authors coding style :) )

#### CHANGELOG.md entry:

[enhancement] `ordered-imports` now supports a groups option to provide custom grouping rules.

palantirtech commented:
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @abierbaum! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

abierbaum commented:
@johnwiseheart @giladgray et al: Is there anything more I need to add before review?

michaeljota commented:
Please see #4141 before you contribute to this project. Thank you!

giladgray commented on src/rules/orderedImportsRule.ts:
```diff
@@ -127,64 +170,105 @@ export class Rule extends Lint.Rules.AbstractRule {
     groupedImports: boolean;
     importSourcesOrderTransform: Transform;
-    moduleSourcePath: Transform;
+    moduleSourceTransform: Transform;
```
😕 changing an option name is often a breaking change... is this exposed to users?

giladgray replied:
looks like it is not exposed, just confirming.

abierbaum replied:
Correct.  This is an internal options object.  I only changed the name to bring parallelism to the "Transform" properties in the interface.

giladgray commented on src/rules/orderedImportsRule.ts:
```diff
@@ -127,64 +170,105 @@ export class Rule extends Lint.Rules.AbstractRule {
+
+    const defaultGroups: JsonGroupOption[] = [
+        { name: "parent directories", match: "^[.][.]", order: 20 },
+        { name: "current directory", match: "^[.]", order: 30 },
```
prefer `\.` over `[.]` - more obvious

abierbaum replied:
I agree but as far as I can tell that isn't allowed in TS/JS inside a string passed to the RegExp.

For example:

```
> RegExp('^\.').test('foo')
> true
> RegExp('^[.]').test('foo')
> false
```
Looks like Javascript will process the "\." escape character when assigning the string and just turn it into "." before it get's to the regex.

Other tools I saw that tried to do similar sorting ended up configuring it with [.][.] as well.

giladgray replied:
ah good point on the string. how about `\\.`? i guess that's the same length, but `[.]` relies on the weird truth that `.` in a character class is treated as the literal `.` character, not the "match-anything" that it means in other contexts.

giladgray commented on src/rules/orderedImportsRule.ts:
```diff
@@ -127,64 +170,105 @@ export class Rule extends Lint.Rules.AbstractRule {
+
+interface JsonGroupOption {
+    name?: string;
+    match: string | null;
```
`null` is not documented above. is it valid in the options?

giladgray replied:
i might prefer not supporting null and expecting a `.*` match-everything pattern. bit of an API question there.

abierbaum replied:
I could see that.  The reason I allowed match to null originally was an implementation detail that could probably be avoided now.  I will change it and see how it works out.

abierbaum replied:
It worked out well so I removed it.  I still use an defaultGroup object internally for unmatched values, but now I use match of `/.*/` there.

giladgray commented on src/rules/orderedImportsRule.ts:
```diff
@@ -127,64 +170,105 @@ export class Rule extends Lint.Rules.AbstractRule {
         "named-imports-order": named = "case-insensitive",
         "module-source-path": path = "full",
-    } = optionSet === undefined ? {} : optionSet;
+        groups: groups = defaultGroups
```
no need for `: groups` since the name is the same. please remove

abierbaum replied:
done.  good catch.

giladgray commented on src/rules/orderedImportsRule.ts:
```diff
@@ -299,49 +403,90 @@ class Walker extends Lint.AbstractWalker<Options> {
     }
 
+    /**
+     * Check all import blogs stopping at the first failure.
```
import blocks?

abierbaum replied:
fixed.

abierbaum commented on src/rules/orderedImportsRule.ts:
```diff
@@ -42,7 +42,7 @@ export class Rule extends Lint.Rules.AbstractRule {
             - Groups of imports are delineated by blank lines. You can use these to group imports
                 however you like, e.g. by first- vs. third-party or thematically or can you can
-                enforce a grouping of third-party, parent directories and the current directory.`,
+                enforce a group order and content.`,
```
I have cleaned up this language:

Groups of imports are delineated by blank lines. You can use this rule to group
imports however you like, e.g. by first- vs. third-party or thematically or
you can define groups based upon patterns in import path names.`

abierbaum commented on src/rules/orderedImportsRule.ts:
```diff
@@ -86,39 +109,59 @@ export class Rule extends Lint.Rules.AbstractRule {
+                            match: { type: "string" },
+                            order: { type: "number" }
+                        },
+                        required: ["match", "order"]
```
I updated it to support passing a string[] or an Object[].

abierbaum commented:
@giladgray I think I addressed all your comments and also added support for the items you suggested.  Thanks for taking the time to review this PR.  Please let me know if you see anything else.

giladgray commented:
@abierbaum please do not amend and force push commits that have already been reviewed. it ruins the diff and makes it hard to incrementally review. we squash everything to one commit when merging anyways, so the default branch is always clean.

giladgray commented:
destroying history shouldn't be part of your default workflow 😄

abierbaum commented:
@giladgray Any more changes need or is this good to go?

giladgray commented:
i would prefer to use `\\.` in the regular expressions -- less _suggestive_

abierbaum commented:
@giladgray Anything else I can do to help get this finished off and merged?

abierbaum commented:
@giladgray Just checking in again on status of this PR.  Our team is excited to get using it after it gets merged.

abierbaum commented:
@giladgray We have been using this now for a couple weeks in production with no issues.  Do you think this could get merged in so it can make the next release?

AlexNisnevich commented:
Any progress on getting this PR merged? It would really help my team's `tslint` workflow.

abierbaum commented:
I don't know of anything left to do.  We have been using it in production with a custom build of tslint for a month.  I would like to get back to mainline though.  Just waiting to hear from @giladgray or @johnwiseheart on acceptability of the PR.

abierbaum commented:
@giladgray and @johnwiseheart .  Just checking in again on the status of this PR.  If this can't land into tslint, let me know so I can put it into a separate extension project like vrsource-tslint-rules

abierbaum commented:
@JoshuaKGoldberg I just fixed up the branch on top of the latest tslint:master.  I had to rerun prettier to remove some merge conflicts.  Anything else you need?

JoshuaKGoldberg commented:
@abierbaum great, nothing else from me!

Other TSLint maintainers _(read: with more experience maintaining)_ have been swamped lately so there's work going on trying to find more folks. I'm holding off from merging larger PRs for a while to give them a chance to either find a full-time maintainer or get to this themselves.

I know that's not a particularly satisfying answer, and neither is the suggested workaround of publishing this a a standalone rule in your own npm package. But expect more soon!

abierbaum commented:
@JoshuaKGoldberg Ok.  Thanks for the update.  I know there are a number of us hoping to get some of these larger improvements through so we can start using them more widely in production.

abierbaum commented:
@JoshuaKGoldberg Any progress on finding more folks so we can get things like this merged?

JoshuaKGoldberg commented:
Sorry, no ☹

nostalgiaz commented:
Any progress?

abierbaum commented:
@nostalgiaz Unfortunately no.  We have been using a build based upon this branch in production for a couple months but we haven't been able to get it merged into the main tslint.  I will probably rebase it again later this week but I was really hoping to get it into the release.

JoshuaKGoldberg commented:
✨ _Progress!_ ✨

A few TSLint maintainers are active now and going through the issues queue. Expect greatness... soon!™

glen-84 commented:
Does this close #1865 as well?

abierbaum commented:
@scote Still waiting for a reviewer.  We have been using it in production for months now and it works great.  Was really hoping it would have gotten in right away.  Looking back I should have done this in an extensions module so more people could get it sooner but I have been trying to add rules to the core instead so people don't need to add extensions.

Koslun commented:
@abierbaum looks to me that the changes have been approved by a reviewer with write access, isn't it now stuck due to the conflicting files?

Or have I misunderstood the process?

abierbaum commented:
@Koslun I could be wrong too.  I was going based upon the tag on it.  Maybe @JoshuaKGoldberg could tell us a bit more.  What do we need to do to get this through?  (I can rebase it when needed to get rid of the conflicts, just didn't want to keep rebasing it multiple times if there hasn't been the review)

JoshuaKGoldberg commented:
Hi!

Although I theoretically can merge this as @Koslun noted, I'm holding off because this is a big enough change that it should be reviewed by someone in @palantirtech. There wasn't anyone available from that group for much of 2018 so the list of PRs that match that description is unfortunately pretty large. There are a few maintainers that are starting to be given time now, so we're just waiting on one of them to get to this one.

scote commented:
Thank you for the update. I would really like to use this new feature.

yanckst commented:
+1

yanckst commented:
Can you resolve the conflicts please?

abierbaum commented:
@yanckst I will take a look this weekend.  It looks like someone else made changes in orderedImportsRule.ts and got it in.  Depending upon how involved those changes are it may take a bit for me to get the conflicts resolved since I had to refactor that file a bit to make it handle the more involved order import sorting this PR adds.

yanckst commented:
@abierbaum Yeah no problem. Take your time and by the way good job!

abierbaum commented:
@JoshuaKGoldberg Is there anyone available to review this so we can get it through.  

I am happy to keep helping on it, but the merge conflicts are a pain to handle when other things get merged to this rule.  The code changes refactor the rule quite a bit to simplify the logic and build everything off a common concept of defined groups.  So any changes to master don't really apply well anymore. :(

abierbaum commented:
@JoshuaKGoldberg Just checking in again to see if there is anyone around to review this.  It looks like there are quite a few people hoping it can get in soon.

adidahiya commented:
@abierbaum sorry for the delay here and thank you for your continued work to keep this PR up to date. I'll make sure to look at this before merging any other changes to this code.

jukben commented:
Haha! This is some stellar work here 🚀 . I'm quite ashamed because I forgot to check pending PRs for such a functionality – so I've hacked my own solution for the very same issue we want to enforce in a company I'm working for.

https://github.com/productboardlabs/tslint-pb#import-group-order ([configuration example](https://github.com/productboardlabs/tslint-pb/blob/master/test/rules/import-group-order/tslint.json#L6), tests are next to it). 

From the what I've seen so far, it's basically the same – this is actually a bit better because you can set how to sort imports withtin the brackets. Which is nice and I don't support it. Just to be sure – this has autofix still enabled for everything, right? Because that's crucial.

Thank you for this, for what I see, once this will make it into the master I can deprecated the solution I wrote without any hassle.

👏

abierbaum commented:
@jukben Yes, it has auto-fix enabled.  That is the entire reason I made the mods.  :)

It works like magic with vscode and typescript auto imports.  I never even go up to my imports anymore.  They just add themselves and format automatically.

abierbaum commented on test/rules/ordered-imports/grouped-imports/test.ts.lint:
```diff
@@ -3,17 +3,17 @@
 
 import {foo, afoo} from 'foo';
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources of different groups must be sorted by: libraries, parent directories, current directory.]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources of different groups must be sorted by: libraries, parent directories, current directory .]
```
@adidahiya The reason for this space came up in the first review.  The idea is that if the user configures a group and uses a regex for `match` but doesn't include a `name`, then the name is the regex.  In that case when the order is printed out the user can get confused because the period will show up for the sentence and look like it is part of the regex.  I will see if I can come up with a better error string and automatic name that is more clear.

abierbaum replied:
@adidahiya I just pushed up a commit that changes this to remove the space and make the default naming for a group with only a regex be /regex/.

So for example from the tests in `groups-string-list`:

```json
{
    "rules": {
        "ordered-imports": [
            true,
            {
                "import-sources-order": "case-insensitive",
                "named-imports-order": "case-insensitive",
                "grouped-imports": true,
                "groups": ["^app", "^@pkg", "^\\.\\.", "^\\."]
            }
        ]
    }
}
```

you end up with a warning like:

```typescript
import { x } from "@pkg/foo";
import { app_f } from "app/foo";
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Imports from this modules are not allowed in this group.  The expected groups (in order) are: /^app/, /^@pkg/, /^\.\./, /^\./.]
```

I think this makes it much more clear that you specified a regex and it is just giving it a default name.  I debated using a different delimiter than `/`, but in the end it seems like this is probably the most familiar to a TS developer as "hey, this is a regex".

Let me know what you think.

abierbaum commented on test/rules/ordered-imports/groups-shared-order/test.ts.lint:
```diff
@@ -0,0 +1,18 @@
+
+import {b} from '@pkg/bar';
+~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources of different groups must be sorted by: ^app_b, ^app_a, ^@pkg .]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources with same group order must be grouped together.]
```
Took me a while to figure it out too and maybe the logic should/could be changed to make this one error.  The issue is that the rule is finding two problems here:  First, the import isn't with the other imports that have the same prefix.  (that is the error about same group order must be grouped together.  Second, if the group was together it would be out of order because this new group has the same order as the previous one.  

Long story short, I think this should probably be just the error about groups must be grouped together.  I will see if I can refactor the logic to make that happen.

abierbaum replied:
@adidahiya Addressed with the change to remove the "grouped together" check.

abierbaum commented on test/rules/ordered-imports/groups-complex/test.ts.lint:
```diff
@@ -0,0 +1,31 @@
+// comment outside of imports
+import {app_f} from 'app/foo';
+import {x} from '@pkg/foo';
+~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources of different groups must be sorted by: extra, ^app, ^@pkg, current dir, parent_dir .]
```
I like your message much better.  I changed it and removed the trailing space.

abierbaum commented on test/rules/ordered-imports/groups-complex/test.ts.lint:
```diff
@@ -0,0 +1,31 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources within a group must be alphabetized.]
+
+import {app_b} from 'app/bar';
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources with same group order must be grouped together.]
```
@adidahiya I think your observation here is key.  From the user's perspective, the first time there is a block out of order it is out of order.  It doesn't matter if that is because the group it should be in has already been seen or if the group they are adding isn't the next group up.  In both cases it is simply out of order.

Once I started reviewing how the code works and what is presented to the user I added a pretty fundamental change the completely removes the "must be grouped together" failure check.  Instead the code now simply looks to see if groups are in order and if they aren't it adds a failure at the first block that is out of order and stops there.

This not only makes the logic in the code much more clear but as you observed this makes the user experience better.

The idea is that once the first block of imports is found that has an import that is out of order, then all future blocks also have blocks out of order.  So instead of flagging all blocks we just flag the first block that is causing the issue and provide a fix that will fix all blocks.

This makes the case for manually fixing the import ordering much more clear because people can solve one item at a time.  After they fix the first item that is out of order, then they move on to fixing the next one and so on.  It also helps to keep the autofix code more simple because the first time we see a failure it returns a fix for the entire set of import blocks.

I put this change on it's own commit so you can easily review the changes this causes for the linting output.  Let me know if you think I should make it work a different way.  I think this works very well, but maybe you have a better idea to try.

abierbaum commented on test/rules/ordered-imports/groups-unmatched/test.ts.lint:
```diff
@@ -0,0 +1,20 @@
+
+import {app_b} from 'app/bar';
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources of different groups must be sorted by: ^app, ^@pkg .]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Import sources with same group order must be grouped together.]
```
Addressed with the change to remove the "grouped together" check.

abierbaum commented:
@adidahiya I think I am ready for you to take a look at the changes I made in response to your review.  Thank you for the great comments.

As far as running this change on the tslint codebase, I am more than happy to do that but I must admit I don't know how to run tslint with fix across the tslint codebase.  If you can point me in the right direction I can submit a PR after this change gets through.

adidahiya commented:
@abierbaum the `lint:from-bin` NPM script (also run as part of `lint`) will run the current tslint source to lint itself, using the root `tslint.json` file. So ordered-imports is already being run that way, but `tslint:all` (src/configs/all.ts) doesn't currently have the grouped imports option enabled -- that's what you'd have to enable.

adidahiya commented:
@abierbaum yeah, I realized that the open PR (#4420) had no conflicts so it was ready to go as-is. thanks for the PR

scote commented:
Do you know when it gonna be release?

abierbaum commented:
@jukben @scote Thanks for the feedback.  I am very interested in a release as well.  Our team is currently running an old version of tslint from a branch with this change.  I am guessing @adidahiya or @JoshuaKGoldberg are the right people to ask about next release timeframes.

RoberMac commented:
cannot wait, please release it! @adidahiya

megabayt commented:
Waiting for release

abierbaum commented:
@adidahiya Any news on a release date or is there a way to use latest HEAD from npm so people can start using this update?

jukben commented:
[OT] @abierbaum yep, you can use git repo url with NPM, Yarn. But I'd like to see release as well. Although I'm afraid it's better to explore ways how to lint TS with ESLint, which going to be a more preferred way in the future.

abierbaum commented:
@jukben I am hoping they do another release before holding up for moving to ESLint.  I don't have the time right now to try to learn a new system and port the rules our team relies up. :(

As far as using git repo url, I haven't found a way to do that with tslint.  When I have tried the packages are not built correctly for use.  Have you had luck doing that?  If so I would *love* to know how.

jukben commented:
Sorry @abierbaum you are right, there has to be build process, my bad. Yeah, I hope for the same, I'm waiting for release as well. I might end up forking this...

Koslun commented:
Concerning the TSLint to ESLint migration I'd like to add that ESLint still doesn't support some of the concepts present in TSLint and that many rules are still missing, including external ones. So even you have the time it wouldn't make sense to completely migrate over to ESLint now as you'd potentially lose many of the rules in-use.

An example is the Angular linting package Codelyzer which is still in TSLint and will only be ported once some TSLint features are merged into ESLint. So given that it may take at least a few months to get ESLint all the missing features of TSLint it will then at least take some more months for the rest of the linting packages like Codelyzer to also migrate to ESLint.

So would be nice to just have this one final release before the inevitable long wait before we are able to actually migrate to ESLint without having regressions.

adidahiya commented:
We are still committed to supporting TSLint for a few months and will continue to do releases to push out PRs like this. "Migrating to ESLint without having any regressions" is of concern to the TSLint community and all of Palantir's TypeScript codebases as well; we want to make that process as smooth as possible. I started this new issue thread to answer questions and clear up doubts about the deprecation path / migration path to ESLint: https://github.com/palantir/tslint/issues/4534. Thanks for your patience as I work to get this PR released.

adidahiya commented:
I just released v5.13.0, please try it out and open new issues if you have problems with this feature. thanks again @abierbaum!

JWess commented:
1) Is there any way to target imports with no "from" such as:

import 'zone.js/dist/zone-testing';

I would like to put them in their own group.

2) Is there any way to specify a max line length for imports so that they will wrap?

Before:
```
import { ExtendSessionDialogComponent, SecondImport } from './core/components/extend-session-dialog/extend-session-dialog.component';
```

After:
```
import { 
  ExtendSessionDialogComponent, SecondImport
} from './core/components/extend-session-dialog/extend-session-dialog.component';
```

JWess commented:
@abierbaum Awesome job! I have a couple questions:

1) Is there any way to target imports with no "from" such as:

import 'zone.js/dist/zone-testing';

I would like to put them in their own group.

2) Is there any way to specify a max line length for imports so that they will wrap?

Before:
```
import { ExtendSessionDialogComponent, SecondImport } from './core/components/extend-session-dialog/extend-session-dialog.component';
```

After:
```
import { 
  ExtendSessionDialogComponent, SecondImport
} from './core/components/extend-session-dialog/extend-session-dialog.component';
```

adidahiya commented:
@JWess please start a new issue for additional features / enhancements to this rule. for (2), that's the job a formatter and such functionality won't be added to tslint at this point (it's not easy, I recommend using prettier instead, see #4534)

gogoyqj commented:
@abierbaum it'll be great if those features supported in typescript-eslint
